### PR TITLE
Add DNSRecord extension resource and controller

### PR DESCRIPTION
**Pull request by stoyanr**
**How to categorize this PR?**

/area control-plane
/kind api-change
/kind enhancement

**What this PR does / why we need it**:
Adds a new `DNSRecord` extension resource for DNS records needed by Gardener, as well as its generic controller, reconciler, actuator, and mapper.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: extensions/pkg/controller/dnsrecord/reconciler.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/controllerutils"
+)
+
+// RequeueAfter is the duration to requeue a dnsrecord reconciliation if indicated by the actuator.
+const RequeueAfter = 2 * time.Second
```
Comment (timebertt) [inline extensions/pkg/controller/dnsrecord/reconciler.go]: This is unused. Do you plan on using it?

File: extensions/pkg/controller/dnsrecord/reconciler.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+func (r *reconciler) reconcile(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, cluster *extensionscontroller.Cluster, operationType gardencorev1beta1.LastOperationType) (reconcile.Result, error) {
+	if err := controllerutils.EnsureFinalizer(ctx, r.reader, r.client, dns, FinalizerName); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.statusUpdater.Processing(ctx, dns, operationType, "Reconciling the dnsrecord"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	r.logger.Info("Starting the reconciliation of dnsrecord", "dnsrecord", dns.Name)
```
Comment (timebertt) [inline extensions/pkg/controller/dnsrecord/reconciler.go]: Add a namespace here?
```suggestion
	r.logger.Info("Starting the reconciliation of dnsrecord", "dnsrecord", client.ObjectKeyFromObject(dns).String())
```

File: extensions/pkg/controller/dnsrecord/reconciler.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+		return reconcile.Result{}, err
+	}
+
+	r.logger.Info("Starting the migration of dnsrecord", "dnsrecord", dns.Name)
+	if err := r.actuator.Migrate(ctx, dns, cluster); err != nil {
+		_ = r.statusUpdater.Error(ctx, dns, extensionscontroller.ReconcileErrCauseOrErr(err), gardencorev1beta1.LastOperationTypeMigrate, "Error migrating dnsrecord")
+		return extensionscontroller.ReconcileErr(err)
+	}
+
+	if err := r.statusUpdater.Success(ctx, dns, gardencorev1beta1.LastOperationTypeMigrate, "Successfully migrated dnsrecord"); err != nil {
```
Comment (timebertt) [inline extensions/pkg/controller/dnsrecord/reconciler.go:171]: Side note, not strictly related to this PR:

I see, this is the same for the other reconcilers and has been like this for a longer time. Still I'm wondering, why the status is set to success, before the finalizers are successfully removed.
I guess, this can be a source a leaked resources (-> seed namespaces) during control plane migration, right?

Should be resolved in a different PR of course, if it's really a problem. I just noticed it during the review here, so I wanted to bring this up.
/cc @plkokanov @kris94

File: pkg/apis/extensions/v1alpha1/types_dnsrecord.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// DNSRecordSpec is the spec of a DNSRecord resource.
+type DNSRecordSpec struct {
+	// DefaultSpec is a structure containing common fields used by all extension resources.
+	DefaultSpec `json:",inline"`
+	// SecretRef is a reference to a secret that contains the cloud provider specific credentials.
+	SecretRef corev1.SecretReference `json:"secretRef"`
+	// Region is the region of this DNS record. If not specified, the region specified in SecretRef will be used.
+	// If that is also not specified, a certain default region will be used, e.g. us-west-2 for AWS route53.
```
Comment (timebertt) [inline pkg/apis/extensions/v1alpha1/types_dnsrecord.go]: Nit, we should probably not include implementation specifics that quickly become outdated in the generic docs.
E.g. like this?
```suggestion
	// If that is also not specified, the extension controller will use its default region.
```

File: pkg/apis/extensions/v1alpha1/types_dnsrecord.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	// +optional
+	Region *string `json:"region,omitempty"`
+	// Zone is the DNS hosted zone of this DNS record. If not specified, it will be determined automatically by
+	// getting all hosted zones of the account and searching for the longest zone name that is a suffix of Name.
+	// +optional
+	Zone *string `json:"zone,omitempty"`
+	// Name is the fully qualified domain name, e.g. api.<shoot domain>.
+	Name string `json:"name"`
```
Comment (timebertt) [inline pkg/apis/extensions/v1alpha1/types_dnsrecord.go]: Q: does the name value have to end in `.` (trailing dot)?
To minimize ambiguity, you could add quotes around the example value.

**Review by timebertt — COMMENTED** [review]
Well done, very clean.
I have only a few minor nits for this PR. But for the POC, this is obviously very good already :)

File: extensions/pkg/controller/dnsrecord/reconciler.go (hunk 3 above)
Comment (stoyanr) [reply extensions/pkg/controller/dnsrecord/reconciler.go:171]: One reason to keep this behavior would be that once the finalizers are removed, if the object is already deleted it immediately disappears and the status is not updateable any longer, the attempt to update it will just result in an error.

> I guess, this can be a source a leaked resources (-> seed namespaces) during control plane migration, right?

If the finalizer is not removed successfully, then deleting the shoot namespace will hang for some time. As far as I can tell there are no other consequences so this should not be critical. For comparison, in the bad case scenario we will leave entire shoot control planes hanging around unused until they are cleaned up properly (manually at first, automatically later). To close the gap, we will have to implement the forcefull "purge" of such control planes at some point anyway.

In short, I think this is rather minor, although of course it could be fixed. I would prefer to keep it as it is though for consistency with other methods (e.g. `delete`).

File: extensions/pkg/controller/dnsrecord/reconciler.go (hunk 1 above)
Comment (stoyanr) [reply extensions/pkg/controller/dnsrecord/reconciler.go]: Removed.

File: extensions/pkg/controller/dnsrecord/reconciler.go (hunk 2 above)
Comment (stoyanr) [reply extensions/pkg/controller/dnsrecord/reconciler.go]: Used `kutil.ObjectName` everywhere, used lowercase `dnsrecord` (as opposed to `DNSRecord`) in all messages, and also improved the phrasing of a few messages. If you like it, I could make these changes across all extension reconcilers in another PR (they are suffering from the same minor bugs / inconsistencies).

File: pkg/apis/extensions/v1alpha1/types_dnsrecord.go (hunk 4 above)
Comment (stoyanr) [reply pkg/apis/extensions/v1alpha1/types_dnsrecord.go]: Done

File: pkg/apis/extensions/v1alpha1/types_dnsrecord.go (hunk 5 above)
Comment (stoyanr) [reply pkg/apis/extensions/v1alpha1/types_dnsrecord.go]: Use quotes as suggested.

**Comment by stoyanr** [comment]
@timebertt Thanks for your suggestions! Addressed them in a new commit.
/squash

**Review by rfranzke — COMMENTED** [review]
Should https://github.com/gardener/gardener/blob/master/docs/extensions/dns.md be adapted with this PR?

File: extensions/pkg/controller/dnsrecord/reconciler.go (hunk 3 above)
Comment (timebertt) [reply extensions/pkg/controller/dnsrecord/reconciler.go:171]: Thanks for the detailed thoughts.

> To close the gap, we will have to implement the forcefull "purge" of such control planes at some point anyway.

Makes sense, got it :)

File: extensions/pkg/controller/dnsrecord/reconciler.go [hunk 6]
```diff
@@ -121,7 +118,7 @@ func (r *reconciler) reconcile(ctx context.Context, dns *extensionsv1alpha1.DNSR
 		return reconcile.Result{}, err
 	}
 
-	r.logger.Info("Starting the reconciliation of dnsrecord", "dnsrecord", dns.Name)
+	r.logger.Info("Starting the reconciliation of dnsrecord", "dnsrecord", kutil.ObjectName(dns))
```
Comment (timebertt) [inline extensions/pkg/controller/dnsrecord/reconciler.go:121]: > Used `kutil.ObjectName` everywhere, used lowercase `dnsrecord` (as opposed to `DNSRecord`) in all messages, and also improved the phrasing of a few messages. If you like it, I could make these changes across all extension reconcilers in another PR (they are suffering from the same minor bugs / inconsistencies).

Yes, I like it very much and it definitely makes sense to make such cleanup/improvement in all (extension) controllers.

If you will do it, I would propose to construct resource-specific loggers first via `.WithValues()` and then pass it around, so one doesn't have to remember to add the resource name to every log statement.

**Review by timebertt — COMMENTED** [review]
No further feedback for this PR in the current stage from my side.
Thanks for incorporating my suggestions!

File: extensions/pkg/controller/dnsrecord/reconciler.go (hunk 3 above)
Comment (plkokanov) [reply extensions/pkg/controller/dnsrecord/reconciler.go:171]: The current flow might cause a bigger problem, although I think the chances of it happening are pretty slim.
During the migration phase, the gardenlet will annotate resources with the migrate annotation and wait for them to be migrated successfully. After that it tries to delete the resources. So if it happened that the status is updted and shows a successful migrate operation, but the finalzers do not get removed due to an error, we actually run the risk of deleting infra resources.

File: extensions/pkg/controller/dnsrecord/reconciler.go (hunk 6 above)
Comment (stoyanr) [reply extensions/pkg/controller/dnsrecord/reconciler.go:121]: OK, I agree, will note it down and open a PR at some point.

**Comment by stoyanr** [comment]
@rfranzke

> Should https://github.com/gardener/gardener/blob/master/docs/extensions/dns.md be adapted with this PR?

I don't think so, since this contract is still there until we have "flipped the switch" with the feature gate introduced in #4253. Instead, I would suggest to add a new markdown file `dnsrecord.md` describing the new contract and mentioning that it's still not used by gardener but will be used in the future.

File: pkg/apis/extensions/v1alpha1/types_dnsrecord.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+const DNSRecordResource = "DNSRecord"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Namespaced,path=DNSRecords,shortName=cp,singular=DNSRecord
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name=Type,JSONPath=".spec.type",type=string,description="The control plane type."
+// +kubebuilder:printcolumn:name=Purpose,JSONPath=".spec.purpose",type=string,description="Purpose of control plane resource."
+// +kubebuilder:printcolumn:name=Status,JSONPath=".status.lastOperation.state",type=string,description="Status of control plane resource."
+// +kubebuilder:printcolumn:name=Age,JSONPath=".metadata.creationTimestamp",type=date,description="creation timestamp"
```
Comment (plkokanov) [inline pkg/apis/extensions/v1alpha1/types_dnsrecord.go]: Hmm, some of the kubebuilder settings seem to be for the controlplane resource.

File: pkg/apis/extensions/v1alpha1/types_dnsrecord.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+)
+
+var _ Object = (*DNSRecord)(nil)
+
+// DNSRecordResource is a constant for the name of the DNSRecord resource.
+const DNSRecordResource = "DNSRecord"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Namespaced,path=DNSRecords,shortName=cp,singular=DNSRecord
```
Comment (timebertt) [inline pkg/apis/extensions/v1alpha1/types_dnsrecord.go]: I think its supposed to be used like this by looking at the generated CRD in `dev/extension-crds`.
```suggestion
// +kubebuilder:resource:scope=Namespaced,path=dnsrecords,shortName=dns,singular=dnsrecord
```

File: pkg/apis/extensions/v1alpha1/types_dnsrecord.go (hunk 7 above)
Comment (stoyanr) [reply pkg/apis/extensions/v1alpha1/types_dnsrecord.go]: Thanks, fixed.

File: pkg/apis/extensions/v1alpha1/types_dnsrecord.go (hunk 8 above)
Comment (stoyanr) [reply pkg/apis/extensions/v1alpha1/types_dnsrecord.go]: Thanks, fixed.

**Review by timebertt — COMMENTED** [review]
Looks great!

File: extensions/pkg/controller/dnsrecord/reconciler.go (hunk 3 above)
Comment (stoyanr) [reply extensions/pkg/controller/dnsrecord/reconciler.go:171]: @plkokanov Looking at https://github.com/stoyanr/gardener/blob/150dc9ace4159945b1de662d14e4abaa992760e3/extensions/pkg/controller/dnsrecord/reconciler.go#L99-L100, I don't think there is a risk to delete the resources (migrated resources are simply skipped by the reconciler), unless the handling is different in other reconcilers (but I don't think so, I see the same check everywhere except for backupbuckets, bastions, and csimigrations). Could I be missing something?

**Comment by stoyanr** [comment]
I added the missing pieces to make this PR mergeable (in separate commits to make reviewing easier):

* Validation for `DNSRecord` resources
* Support for deletion protection of `DNSRecord` resources
* Documentation about the `DNSRecord` contract (as requested above)
* Unit tests

I consider this PR now ready to be reviewed and merged if there are no further comments.
/cc @timebertt @rfranzke @plkokanov

File: charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml [hunk 9]
```diff
@@ -0,0 +1,281 @@
+
```
Comment (timebertt) [inline charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml]:
```suggestion
```

File: pkg/apis/extensions/validation/dnsrecord.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+	if len(spec.Values) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("values"), "field is required"))
+	}
+	if spec.RecordType == extensionsv1alpha1.DNSRecordTypeCNAME && len(spec.Values) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("values"), spec.Values, "CNAME records must have a single value"))
+	}
+	for _, value := range spec.Values {
+		allErrs = append(allErrs, validateValue(spec.RecordType, value, fldPath.Child("values"))...)
+	}
```
Comment (timebertt) [inline pkg/apis/extensions/validation/dnsrecord.go]:
```suggestion
	for i, value := range spec.Values {
		allErrs = append(allErrs, validateValue(spec.RecordType, value, fldPath.Child("values").Index(i))...)
	}
```

File: pkg/apis/extensions/validation/dnsrecord_test.go [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+				"Field": Equal("spec.recordType"),
+			})), PointTo(MatchFields(IgnoreExtras, Fields{
+				"Type":  Equal(field.ErrorTypeRequired),
+				"Field": Equal("spec.values"),
+			}))))
+		})
+
+		It("should forbid non-nil but empty region or zone", func() {
+			dns.Spec.Region = pointer.StringPtr("")
+			dns.Spec.Zone = pointer.StringPtr("")
```
Comment (timebertt) [inline pkg/apis/extensions/validation/dnsrecord_test.go]: Use `pointer.String` instead, see https://github.com/gardener/gardener/pull/4292

File: charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.6.0
```
Comment (timebertt) [inline charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml]: We can remove the annotation

File: charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+  creationTimestamp: null
```
Comment (timebertt) [inline charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml]: Please add the following label, otherwise the CRD will not be deletion protected
```suggestion
  labels:
    gardener.cloud/deletion-protected: "true"
```

because there is an object selector in the GSAC webhook config:
https://github.com/gardener/gardener/blob/712a875b5965c9c87a8ce9c304c13733fe4b11ce/pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go#L313-L315

We should add the same object select to the webhook config in the envtest, so that it can catch such issues: https://github.com/gardener/gardener/blob/ce60b9be18f869e0fa7917c91526072d0ca66a1e/pkg/seedadmissioncontroller/seedadmission_suite_test.go#L113
(or even better: reuse the webhook config from `pkg/operation/botanist/component/seedadmissioncontroller`)

File: charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+    served: true
+    storage: true
+    subresources:
+      status: {}
+status:
+  acceptedNames:
+    kind: ""
+    plural: ""
+  conditions: []
+  storedVersions: []
```
Comment (timebertt) [inline charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml]:
```suggestion
```

File: docs/extensions/dnsrecord.md [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+* The fully qualified domain name (FQDN) of the DNS record, e.g. "api.\<shoot domain\>".
+* The DNS record type, one of `A`, `CNAME`, or `TXT`.
+* The DNS record values, that is a list of IP addresses for A records, a single hostname for CNAME records, or a list of texts for TXT records.
+
+Optionally, the `DNSRecord` resource may contain also the following information:
+
+* The region of the DNS record. If not specified, the region specified in the referenced `Secret` shall be used. If that is also not specified, the extension controller shall use a certain default region. 
+* The DNS hosted zone of the DNS record. If not specified, it shall be determined automatically by the extension controller by getting all hosted zones of the account and searching for the longest zone name that is a suffix of the fully qualified domain name (FQDN) mentioned above. 
+* The TTL of the DNS record in seconds. If not specified, it shall be set by the extension controller to 120.
+
```
Comment (timebertt) [inline docs/extensions/dnsrecord.md:52]:
```suggestion

**Example `DNSRecord`**:
```

**Review by timebertt — COMMENTED** [review]
Well done! I like the documentation / contract description.
This is looking very good to me. 
Only a few minor things left.

File: docs/extensions/dnsrecord.md (hunk 15 above)
Comment (stoyanr) [reply docs/extensions/dnsrecord.md:52]: Done

File: charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml (hunk 14 above)
Comment (stoyanr) [reply charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml]: Done

File: charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml (hunk 13 above)
Comment (stoyanr) [reply charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml]: Added the label, I believe the rest is not for this PR.

File: charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml (hunk 12 above)
Comment (stoyanr) [reply charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml]: Done

File: pkg/apis/extensions/validation/dnsrecord_test.go (hunk 11 above)
Comment (stoyanr) [reply pkg/apis/extensions/validation/dnsrecord_test.go]: Done, I should adapt the unit tests in #4253 as well.

File: pkg/apis/extensions/validation/dnsrecord.go (hunk 10 above)
Comment (stoyanr) [reply pkg/apis/extensions/validation/dnsrecord.go]: Done

File: charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml (hunk 9 above)
Comment (stoyanr) [reply charts/seed-bootstrap/charts/extensions/templates/crd-dnsrecord.yaml]: Done

**Review by timebertt — APPROVED** [review]
Nice!
/lgtm
